File: conformance/tests/token-nft/token-nft.runar.zig
const runar = @import("runar");

pub const SimpleNFT = struct {
    pub const Contract = runar.StatefulSmartContract;

    owner: runar.PubKey = "000000000000000000000000000000000000000000000000000000000000000000",
    tokenId: runar.ByteString,
    metadata: runar.ByteString,

    pub fn init(owner: runar.PubKey, tokenId: runar.ByteString, metadata: runar.ByteString) SimpleNFT {
        return .{
            .owner = owner,
            .tokenId = tokenId,
            .metadata = metadata,
        };
    }

    pub fn transfer(self: *SimpleNFT, ctx: runar.StatefulContext, sig: runar.Sig, newOwner: runar.PubKey, outputSatoshis: i64) void {
        runar.assert(runar.checkSig(sig, self.owner));
        runar.assert(outputSatoshis >= 1);
        ctx.addOutput(outputSatoshis, .{ newOwner });
    }

    pub fn burn(self: *const SimpleNFT, sig: runar.Sig) void {
        runar.assert(runar.checkSig(sig, self.owner));
    }
};

File: conformance/tests/token-nft/token-nft.runar.go
//go:build ignore

package contract

import "runar"

type SimpleNFT struct {
	runar.StatefulSmartContract
	Owner    runar.PubKey
	TokenId  runar.ByteString `runar:"readonly"`
	Metadata runar.ByteString `runar:"readonly"`
}

func (c *SimpleNFT) Transfer(sig runar.Sig, newOwner runar.PubKey, outputSatoshis runar.Bigint) {
	runar.Assert(runar.CheckSig(sig, c.Owner))
	runar.Assert(outputSatoshis >= 1)
	c.AddOutput(outputSatoshis, newOwner)
}

func (c *SimpleNFT) Burn(sig runar.Sig) {
	runar.Assert(runar.CheckSig(sig, c.Owner))
}
